# Домашнє завдання до теми "Основи роботи з даними у Pandas"

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [ ]:
import pandas as pd


In [ ]:
!ls drive/MyDrive/Files/supermarket_sales.csv

drive/MyDrive/Files/supermarket_sales.csv


In [ ]:
df = pd. read_csv ('drive/MyDrive/Files/supermarket_sales.csv')

In [ ]:
type (df)

pandas.core.frame.DataFrame

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [ ]:
df.head (4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4


In [ ]:
df.tail(4)

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.690,1022.490,3/2/2019,17:16,Ewallet,973.80,4.761905,48.690,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.592,33.432,2/9/2019,13:22,Cash,31.84,4.761905,1.592,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.291,69.111,2/22/2019,15:33,Cash,65.82,4.761905,3.291,4.1
999,849-09-3807,A,Yangon,Member,Female,Fashion accessories,88.34,7,30.919,649.299,2/18/2019,13:28,Cash,618.38,4.761905,30.919,6.6


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [ ]:
df[615:620:1]

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
615,509-29-3912,A,Yangon,Member,Female,Sports and travel,38.72,9,17.4240,365.9040,3/20/2019,12:24,Ewallet,348.48,4.761905,17.4240,4.2
616,327-40-9673,B,Mandalay,Member,Male,Sports and travel,72.60,6,21.7800,457.3800,1/13/2019,19:51,Cash,435.60,4.761905,21.7800,6.9
617,840-19-2096,C,Naypyitaw,Member,Male,Electronic accessories,87.91,5,21.9775,461.5275,3/14/2019,18:10,Ewallet,439.55,4.761905,21.9775,4.4
618,828-46-6863,A,Yangon,Member,Male,Food and beverages,98.53,6,29.5590,620.7390,1/23/2019,11:22,Credit card,591.18,4.761905,29.5590,4.0
619,641-96-3695,C,Naypyitaw,Member,Female,Fashion accessories,43.46,6,13.0380,273.7980,2/7/2019,17:55,Ewallet,260.76,4.761905,13.0380,8.5


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [ ]:
df.dtypes

,0
Invoice ID,object
Branch,object
City,object
Customer type,object
Gender,object
Product line,object
Unit price,float64
Quantity,int64
Tax 5%,float64
Total,float64


Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [ ]:
statistics = df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99])
print(statistics)


count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64


Пояснення колегам:

***Якщо ми візьмемо всі дані про дохід і впорядкуємо їх від найменшого до найбільшого, то 75-й персентиль показує певне значення доходу. Це означає, що три чверті (75%) від усіх доходів є меншими або рівними цьому значенню, і лише одна чверть (25%) доходів є вищими за нього***

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [ ]:
df.City.value_counts()

,count
City,
Yangon,340
Mandalay,332
Naypyitaw,328


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [ ]:
city_counts = df.City.value_counts()
city_counts

,count
City,
Yangon,340
Mandalay,332
Naypyitaw,328


In [ ]:
city_counts / city_counts.sum()

,count
City,
Yangon,0.340
Mandalay,0.332
Naypyitaw,0.328


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [ ]:
mean_Yangon = df[df.City=='Yangon']['gross income'].mean()
mean_Mandalay = df[df.City=='Mandalay']['gross income'].mean()

In [ ]:
mean_Yangon, mean_Mandalay

(np.float64(14.874001470588235), np.float64(15.232024096385542))

У місті Yangon середнє значення становить- 14.87, а в місті Mandalay - 15.23
Три гіпотези чому так може бути:
1.У місті Mandalay може бути більша частка клієнтів із високою купівельною спроможністю або вищий середній чек.
2.У місті з вищим середнім доходом краще продаються продукти з вищою маржинальністю або певні категорії товарів, які користуються більшим попитом.
3.У місті Mandalay можуть частіше використовувати швидші або зручніші методи оплати, що пришвидшує обслуговування клієнтів і збільшує пропускну здатність магазину.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [ ]:
df['gross margin percentage'].nunique()

1

***У всіх всіх інвойсах однакова маржа***

Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [ ]:
df[df.City=='Yangon']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Home and lifestyle,0.191176
Electronic accessories,0.176471
Sports and travel,0.173529
Food and beverages,0.170588
Fashion accessories,0.150000
Health and beauty,0.138235


In [ ]:
df[df.City=='Mandalay']['Product line'].value_counts(normalize=True)

,proportion
Product line,
Fashion accessories,0.186747
Sports and travel,0.186747
Electronic accessories,0.165663
Health and beauty,0.159639
Food and beverages,0.150602
Home and lifestyle,0.150602


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)